tasks/change-granule-collection-s3/src/index.ts

```diff
@@ -0,0 +1,433 @@
+    const cmrObject = cmrObjectsByGranuleId[targetGranule.granuleId];
+    if (!(cmrFile && cmrObject)) {
+      outputObjects[targetGranule.granuleId] = {};
+    }
+    outputObjects[targetGranule.granuleId] = updateCmrFileCollections({
+      collection: config.targetCollection,
+      cmrFileName: cmrFile.key,
+      cmrObject,
+      files: (targetGranule as ValidGranuleRecord).files,
+      distEndpoint,
+      bucketTypes,
+      cmrGranuleUrlType,
+      distributionBucketMap,
+    });
+  });
+  return outputObjects;
+}
+
+/**
+ * Build a set of granules according to new collection
+ * New granules reference new collectionId as their collectionId
+ * files for new granules are updated according to new collection url_path
+ * file names are *not* updated
+ */
+function buildTargetGranules(
+  granules: Array<ValidGranuleRecord>,
+  config: EventConfig,
+  cmrObjects: { [key: string]: Object },
+  targetCollection: CollectionRecord
+): Array<ValidGranuleRecord> {
+  const bucketsConfig = new BucketsConfig(config.buckets);
+  const targetGranules: Array<ValidGranuleRecord> = [];
+  const granulesAndMetadata = granules.map(
+    async (granule) => updateGranuleMetadata(
+      granule,
+      bucketsConfig,
+      cmrObjects,
+      targetCollection
+    )
+  );
+  granulesAndMetadata.forEach((targetGranule) => {
+    targetGranules.push(targetGranule);
+  });
+  return targetGranules;
+}
+
+async function changeGranuleCollectionS3(event: ChangeCollectionsS3Event): Promise<Object> {
+  const config = event.config;
+  const s3MultipartChunksizeMb = config.s3MultipartChunksizeMb
+    ? config.s3MultipartChunksizeMb : Number(process.env.default_s3_multipart_chunksize_mb);
+
+  const chunkSize = s3MultipartChunksizeMb ? s3MultipartChunksizeMb * MB : undefined;
+  const targetCollection = await getCollection({
+    prefix: getRequiredEnvVar('stackName'),
+    collectionName: config.targetCollection.name,
+    collectionVersion: config.targetCollection.version,
+  });
+
+  log.debug(`change-granule-collection-s3 config: ${JSON.stringify(event.config)}`);
+
+  const granuleIds = event.input.granuleIds;
+  const tempGranulesInput = await Promise.all(granuleIds.map((granuleId) => getGranule({
+    prefix: getRequiredEnvVar('stackName'),
+    granuleId,
+  })));
+  const invalidBehavior = config.invalidBehavior || 'skip';
+  let granulesInput: Array<ValidGranuleRecord>;
+  if (invalidBehavior === 'skip') {
+    granulesInput = tempGranulesInput.filter((granule) => {
+      if (!apiGranuleRecordIsValid(granule)) {
+        log.warn(`granule has unparseable file details ${granule}`);
+        return false;
+      }
+      return true;
+    }) as ValidGranuleRecord[];
+  } else {
+    tempGranulesInput.forEach((granule) => {
+      if (!apiGranuleRecordIsValid(granule)) {
+        throw new Error(`granule has unparseable file details ${granule}`);
+      }
+    });
+    granulesInput = tempGranulesInput as ValidGranuleRecord[];
```

Comment:
Instead of a forced type assertion, use


```suggestion
    granulesInput = tempGranulesInput.filter(apiGranuleRecordIsValid);
```

Reply:
the previous block *does* guarantee that all of those are ValidGranuleRecord(s)

Reply:
Until the block is changed - I'm arguing using >as< is a maintenance hazard over a code typeguard.     Also - what this really appears to be asserting is that there won't be any missing files, etc, due to `Type 'undefined' is not assignable to type 'ValidApiGranuleFile[]'.`

Reply:
This is sorted via huddle